plugins/flatpak/gs-flatpak.c

```diff
@@ -3178,6 +3184,66 @@ gs_flatpak_create_runtime_repo (GsFlatpak *self,
 	return g_steal_pointer (&app);
 }
 
+GsApp *
+gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+				     GFile *file,
+				     GCancellable *cancellable,
+				     GError **error)
+{
+	g_autoptr(GPtrArray) remotes = NULL;
+	g_autofree gchar *uri = NULL;
+	gboolean dir_has_repo = FALSE;
+	gboolean reload_overview = FALSE;
+	GsApp *app = NULL;
+	FlatpakRemoteType usb_type[] = { FLATPAK_REMOTE_TYPE_USB };
+
+	remotes = flatpak_installation_list_remotes_by_type (self->installation,
+							     usb_type,
+							     G_N_ELEMENTS (usb_type),
+							     cancellable, error);
+	if (remotes == NULL)
+		return NULL;
+
+	uri = g_file_get_uri (file);
+
+	for (guint i = 0; i < remotes->len; ++i) {
+		FlatpakRemote *remote = g_ptr_array_index (remotes, i);
+		g_autofree gchar *url = flatpak_remote_get_url (remote);
+		g_autoptr(GError) local_error = NULL;
+
+		/* skip if the remote does not have the dir's path as a prefix */
+		if (!g_str_has_prefix (url, uri))
+			continue;
+
+		dir_has_repo = TRUE;
+
+		if (!gs_flatpak_mark_apps_from_usb_remote (self, remote, &reload_overview,
+							   cancellable, &local_error)) {
+			g_debug ("Failed to mark USB apps from remote at %s: %s",
+				 url, local_error->message);
+			continue;
+		}
+	}
+
+	if (!dir_has_repo) {
+		g_set_error (error, G_IO_ERROR_FAILED, G_IO_ERROR_NOT_SUPPORTED,
```

Comment:
Would `G_IO_ERROR_NOT_FOUND` be a more appropriate error code?

Reply:
It was meant to say that loading a repo from such a dir is not supported (because we don't know the dir's format).
But I don't mind changing it to `NOT_FOUND` since we cannot find a remote in it...